### ISSUE TITLE
Data Source `dynatrace_entities` doesn't set the attribute `entities` if none are matching the specified entity selector

### ISSUE THREAD

**Issue by Reinhard-Pilz-Dynatrace** [dynatrace-oss/terraform-provider-dynatrace#529]
**Describe the bug**
The data source `dynatrace_entities` doesn't set the attribute `entities` in case the REST API responds back with no results to the specified filter.
That behavior results in making the data source unusable in modules where the user wants to loop over the results or wants to just count them. Terraform will throw an error, because the data type (nil) cannot get iterated over.

**Expected behavior**
The attribute `entities` should always get populated. If no entities are matching the filter, then an empty list should be be set for that attribute

**Comment by mattBaumBeneva** [comment]
@kishikawa12 , our tests are failing even with this fix.
For example, when we configure settings on a process group, which may or may not be present on a SaaS tenant:

```
# get PGs by name
data "dynatrace_entities" "process_groups" {
  entity_selector = "type(\"PROCESS_GROUP\"),entityName(\"${var.nom}\"),mzId(\"${local.proprietaire.legacy_id}\")"
}

# configure PG Alerting
resource "dynatrace_pg_alerting" "alerting" {
  count = length(data.dynatrace_entities.process_groups.entities)

  enabled       = var.alertage.disponibilite.actif
  process_group = data.dynatrace_entities.process_groups.entities[count.index].entity_id

  alerting_mode = var.alertage.disponibilite.actif ? "ON_PGI_UNAVAILABILITY" : null
}
```

Terraform fails at the Plan step:
```
12:24:33  │ Error: Invalid count argument
12:24:33  │ 
12:24:33  │   on ../../src/modules/blueprints/Processus/main.tf line 31, in resource "dynatrace_pg_alerting" "alerting":
12:24:33  │   31:   count = length(data.dynatrace_entities.process_groups.entities)
12:24:33  │ 
12:24:33  │ The "count" value depends on resource attributes that cannot be determined
12:24:33  │ until apply, so Terraform cannot predict how many instances will be
12:24:33  │ created. To work around this, use the -target argument to first apply only
12:24:33  │ the resources that the count depends on.
```

Tested with provider 1.67.0 and terraform 1.9.6 (latest).

**Comment by kishikawa12** [comment]
Hi @mattBaumBeneva,

We believe the issue here may be the dependency usage of `local.proprietaire.legacy_id`, are you able to provide details on how that's being retrieved?  Hopefully we can find a solution to handle that in our provider, but we wonder it may be a limitation of Terraform.

Thanks!
Kodai

**Comment by mattBaumBeneva** [comment]
Sure, 2 cases exist:

1. The repo configuring the setting also configures the Management Zone (IDs are output in an "info" object from that module)
2. The repo references a MZ configured elsewhere (we recover the MZ by name)

```
module "mz" {
  source = "../../resources/management_zone"
  nom    = var.proprietaire.nom
}

locals {
  proprietaire = (var.proprietaire.id == "" ? module.mz.info : var.proprietaire)
}
```

In this case, we reference a Team/MZ on which the module depends (all of our working modules do the same thing:

```
module "fake-pg" {
  source = "../../src/modules/blueprints/Processus"

  proprietaire     = module.Equipe-test-01.info
  nom              = "my-fake-pg"

  # the rest of the module ...
}
```

The output in question is defined as follows, but it has always worked this way:

```
output "info" {
  value = {
    nom       = dynatrace_management_zone_v2.mz.name,
    id        = dynatrace_management_zone_v2.mz.id,
    legacy_id = dynatrace_management_zone_v2.mz.legacy_id,
  }
}
```

Thanks.

**Comment by Reinhard-Pilz-Dynatrace** [comment]
Hello Matt,

This looks a bit like a chicken and egg problem, I'm afraid.

In order to decide how many instances of `dynatrace_pg_alerting.alerting` to create, Terraform requires that whatever is specified via `count` meta attribute to be available already after the planning phase.
That's indeed not a problem when working with Data Sources (like in this case `dynatrace_entities`). Terraform evaluates them already during the planning phase.

... unless that Data Source depends on other resources. The ID of the management zone specified within the entity selector
```
entity_selector = "type(\"PROCESS_GROUP\"),entityName(\"${var.nom}\"),mzId(\"${local.proprietaire.legacy_id}\")"
```
cannot be available before the module that creates the management zone has actually gotten applied - the phase afterwards.

A possible solution might be switching to `mzName` instead of `mzId` within the entity selector. The name of the management zone is something your modules should be able to determine already during the planning phase. Perhaps there exists a way to pass along the name of the management zone on its own?

In either case (when using an `mzId` AND when using `mzName`) you need to be aware of an important fact:
The first `terraform apply` won't be creating all the `dynatrace_pg_alerting` instances right away. It will require a subsequent `terraform apply` a later on for that.

Reason for that is, that creating a Management Zone in Dynatrace is a lazy process (both, using the REST API and the WebUI). Creating a Management Zone triggers a background process on your Dynatrace cluster. It iterates over all existing entities, so it can match them with the entity selector and assign the ones that pass the test to that Management Zone.
In a Dynatrace environment with little load and not a lot of components connected that's negligible.
But in a Production Environment you can expect that to take minutes or hours, until ALL matching entities are finally assigned to the newly created Management Zone.

In other words, you cannot expect `dynatrace_entities` to come back with all the entities that belong to a Management Zone if that Mgmz has just been created a few milliseconds before.

Using `mzName` in the entity selector the behavior is similar - and more apparent and predictable.
With that approach we're essentially allowing Terraform to evaluate the Data Source before the Management Zone has even been created. In other words, the first `terraform apply` will DEFINITELY not contain any instances of `dynatrace_pg_alerting.alerting` - the Management Zone doesn't exist yet, therefore no Process Groups will match. Only a subsequent `terraform apply` will create these instances.

Makes sense?

**Comment by mattBaumBeneva** [comment]
Yes, that makes sense. We were already aware of this timing issue requiring multiples Apply steps. We will try switching to mzName.

**Comment by mattBaumBeneva** [comment]
Upon further testing, using mzName does not either allow applying all configs at once (the API resturns an error indicating that no such MZ exists yet).
We have altered our code to refer to the autotags that must be applied to the entities in question.

I understand this also requires multiple Apply runs to fully configure Dynatrace in the case that autotagging also takes time to apply, but at least in this case the circular dependency is gone and the API will not complain about the entity selector in question (tags need not exist to evaluate the selector).

I think we can close this issue. Our build is now passing.

**Comment by Reinhard-Pilz-Dynatrace** [comment]
Thanks for the update!

**Comment by Reinhard-Pilz-Dynatrace** [comment]
Hello @mattBaumBeneva 

I've reopened that ticket because another user seems to have hit a very similar road block.

Let's say `dynatrace_entities` was able to simply produce an empty list for the entries in case the Management Zone doesn't exist (when using `mzName(...)`).
Would you be able to go with the solution I had proposed (decoupling the data source from the applied resources)?
Or would that still be some sort of second choice?

**Comment by mattBaumBeneva** [comment]
Hi, @Reinhard-Pilz-Dynatrace . If the provider can catch the error where the MZ name is not found and return an empty list, I would consider that expected behavior.
From a consumer's perspective, if I define a selector query, I expect one of two results:

1. entities are found (non-empty list)
2. entities are not found (empty list)

In my view, the fact that the MZ was not found should be additional information explaining why the list is empty.

In the case you describe, we could revert our logic to using the mzName. I think our code would be a bit cleaner that way. 

We are currently producing entity selectors on autotag values (instead of the MZ), which are created when the MZ is created. This leaks implementation details from our Team module to other modules (Process Group, Service).

**Comment by Reinhard-Pilz-Dynatrace** [comment]
I actually had to figure out that difference in behavior myself earlier today when experimenting around with the REST endpoint that's being used when querying for entities.

It looks like the REST API doesn't complain when specifying the ID of a non-existent Management Zone, but it returns an error message when specifying the Name of a Management Zone that does (not yet) exist.
IMHO that looks like a shortcoming in the REST API, but one that can get worked around within the Terraform Provider.

Regarding the additional information explaining why the list is empty; I can think of two options here:
* We introduce a computed attribute `errors` for the data source `dynatrace_entities` that contains that error message
* The Data Source `dynatrace_entities` could tell Terraform to print out a warning onto console in that case

I'm open for both approaches. As a Terraform power user I don't mind following your preference.

**Comment by mattBaumBeneva** [comment]
Yes, that's what I noticed as well. The REST API is inconsistent in its behavior.

As to the warning message, I think it would be preferable for Terraform to print it, the same way deprecation warnings are issued.
We run our builds through Jenkins, and these messages appear in color in the GUI and are easily seen in a log.

**Comment by Reinhard-Pilz-Dynatrace** [comment]
Sounds good. This is the output the data source will produce with the release scheduled for tomorrow:
```
You can apply this plan to save these new output values to the Terraform state, without changing any real infrastructure.
╷
│ Warning: Scope: mzname. Management-Zone not found: CatchMeIfYouCan
│
│   with data.dynatrace_entities.foo,
│   on main.tf line 16, in data "dynatrace_entities" "foo":
│   16: data "dynatrace_entities" "foo" {
│
╵
```

**Comment by Reinhard-Pilz-Dynatrace** [comment]
@mattBaumBeneva I'm closing this ticket until further notice. Release v1.68.1 contains the changes we've talked about. Let us know whether it has worked out as expected for you.
